README.md
# AngularJS ES6

> A starting point for building web applications with AngularJS 1.5 using ES6, Webpack and PostCSS.


## Prerequisites

- [Node.js](https://nodejs.org/en/download/).
- [npm](https://www.npmjs.com/), installed with Node.js.
- [yarn](https://yarnpkg.com/) a package manager for project dependencies.


## Install dependencies

#### 1) Check `Node.js` version.
```sh
node --version
```
The version should be at or above 6.9.1

If you don't have Node.js installed go to [nodejs](https://nodejs.org/en/download/) and install the appropiate version or use [nvm](http://www.sergiolepore.net/2014/06/30/nvm-instalando-y-usando-node-version-manager/) (Recommended).

#### 2) Check `yarn` version.
```sh
yarn --version
```
The version should be at or above 0.15.1

If you don't have yarn installed go to [yarn](https://yarnpkg.com/en/docs/install) and install the appropiate version.

#### 3) Install the workspace dependencies.
```sh
# cd to your project folder
yarn
```


## Workflow

### Environment variables

#### Host configuration
- `NODE_ENV="development"`. Environment for server.
- `ANGULARJS_ES6_HOST=localIP`. Host of webapplication.
- `ANGULARJS_ES6_PORT=3000`. Port of webapplication (Use in development).

#### Access URLs
- `ANGULARJS_ES6_API_URL="http://${ip.address()}:${8080}/api"`. Url of API.

To change any of this variables:

```sh
export ANGULARJS_ES6_PORT=8000
```

### Development workflow

#### Static server with live reload
```sh
# cd to your project folder
yarn start
# After this, a message indicate the url to run the application
```
This create a server using `webpack-dev-server` to serve the application in development environment.  
The browser reloads the app when any file change.

### Production workflow
```sh
# cd to your project folder
# Set environment variable for production
export NODE_ENV='production'

# Build
yarn run build
```
Put all in `dist` folder ready for production.

### Install/Uninstall/Update dependencies
Follow this rules to update dependencies:

- Install dependencies

```sh
# To install production dependencies
# cd to your project folder
yarn add angular --exact
```
```sh
# To install development dependencies
# cd to your project folder
yarn add eslint --exact --dev
```

- Uninstall dependencies

```sh
# cd to your project folder
yarn remove angular
```

- **To Upgrade any dependencies follow `Uninstall dependencies` step and install again like `Install dependencies`.**

### Others scripts

To run eslint in console type:

```sh
# cd to your project folder
yarn run lint
```


## Styling

#### CSS Framework **[niduscss-framework](https://github.com/nimedev/niduscss-framework)**.

#### Naming conventions **[SuitCSS naming conventions](https://github.com/suitcss/suit/blob/master/doc/naming-conventions.md)**.


## Angular HTML5 mode or pretty URLs

This mode is set by default. If use it consider this:

- All call related to links like `href="/link/to/anyware"`, `$location.path('/link/to');`, etc. result as a hit to server. Use it only for SEO porpuses.
- To avoid unnecesary hits to server (SPA) you can use `ui-sref="/link/to"` in html files or `$state` service in angular files.


## Editors
This project is configured to use with `VScode`. Also contain pre-configured task for development. See `.vscode` folder to explore the options.


## [Contributing](CONTRIBUTING.md)


## [Changelog](CHANGELOG.md)


## [License](LICENSE.md)

CHANGELOG.md
# Change Log

## [3.0.0] - 2012-11-17

### Modified
- Change folder structure.

### Updated
- Upgrade css-loader.


## [2.4.0] - 2012-11-16

### Added
- Minify index.html when build.

### Modified
- Update vscode settings.

### Updated
- Upgrade autoprefixer, copy-webpack-plugin, eslint, postcss-import, postcss-loader and webpack-merge.


## [2.3.0] - 2012-11-09

### Modified
- Use environment variables to set constants in webpack module.
- Update lint:js script.

### Updated
- Remove typings.
- Upgrade autoprefixer, babel-core, clean-webpack-plugin, eslint-loader, eslint, html-webpack-plugin, postcss-loader, ip and postcss-import.


## [2.2.0] - 2012-10-27

### Modified
- Move some assets to public folder.

### Updated
- Use nodejs v6.9.1.
- Upgrade niduscss-elements-form and niduscss-lib-mixins.
- Upgrade stylelint, babel-core, babel-loader, babel-preset-es2015, copy-webpack-plugin, html-webpack-plugin and webpack.

## [2.1.1] - 2012-10-18

### Updated
- Upgrade eslint and webpack-merge.


## [2.1.0] - 2012-10-17

### Modified
- Import base styles before all components.

### Updated
- Use nodejs v6.8.1.
- Upgrade eslint-loader, postcss-loader and postcss-mixins.


## [2.0.1] - 2012-10-15

### Fixed
- Remove constants.js from git control.


## [2.0.0] - 2012-10-15

### Added
- Simplify folder structure to start.

### Modified
- Migrate to yarn.

### Updated
- autoprefixer clean-webpack-plugin eslint stylelint.
- Use nodejs v6.8.0.


## [1.2.0] - 2012-10-09

### Modified
- Modify environment variables for API.

### Updated
Upgrade clean-webpack-plugin, eslint-config-nimedev and stylelint.


## [1.1.2] - 2012-10-07

### Modified
- Remove root property from resolve en webpack.config file.


## [1.1.1] - 2012-10-07

### Modified
- Improve webpack configuration.


## [1.1.0] - 2012-10-07

### Deleted
- Remove node-notifier package.

### Modified
- Clean niduscss framework.
- Improve styles folder.
- Move app.constants.js to shared folder.
- Move shared folder to src.
- Move typings to src folder.

### Updated
- Upgrade postcss-mixins.


## [1.0.5] - 2012-10-06

### Fixed
- Change application name in package.json file.


## [1.0.4] - 2012-10-06

### Modified
- Export constants through app index. 


## [1.0.3] - 2012-10-06

### Fixed
- Little fixes.


## [1.0.2] - 2012-10-06

### Modified
- Remove const prefix from modules files.


## [1.0.1] - 2012-10-06

### Modified
- Change arguments in class constructors.

### Updated
- Upgrade webpack-dev-server package.


## [1.0.0] - 2012-10-05

* Initial release.

webpack.config.js
// Set default node environment to development
process.env.NODE_ENV = process.env.NODE_ENV || 'development'

const path = require('path')
const HtmlWebpackPlugin = require('html-webpack-plugin')
const ip = require('ip')
const merge = require('webpack-merge')
const opener = require('opener')
const validate = require('webpack-validator')
const webpack = require('webpack')

// Configuration parts
const parts = require('./libs/webpack.parts')

// Constants & Variables
const host = process.env.ANGULARJS_ES6_HOST || ip.address()
const port = process.env.ANGULARJS_ES6_PORT || 3000
const apiUrl = process.env.ANGULARJS_ES6_API_URL || `http://${ip.address()}:${8080}/api`
const PATHS = {
  app: path.join(__dirname, 'app'),
  dist: path.join(__dirname, 'dist'),
  images: path.join(__dirname, 'public/assets/img'),
  fonts: path.join(__dirname, 'public/assets/fonts'),
  copy: [
    {
      from: './public/assets',
      to: 'assets'
    }, {
      from: './public/favicon.ico'
    }, {
      from: './public/robots.txt'
    }]
}

// Common settings for webpack
const common = {
  entry: {
    app: PATHS.app
  },
  output: {
    path: PATHS.dist,
    filename: '[name].js'
  },
  plugins: [
    new webpack.DefinePlugin({
      'process.env': {
        'NODE_ENV': JSON.stringify(process.env.NODE_ENV),
        'ANGULARJS_ES6_API_URL': JSON.stringify(apiUrl)
      }
    }),
    new HtmlWebpackPlugin({
      template: './public/index.html',
      hash: true,
      minify: {
        collapseWhitespace: true,
        removeComments: true
      }
    })
  ],
  resolve: {
    extensions: ['', '.js', '.json', '.css']
  },
  resolveLoader: {
    modulesDirectories: ['node_modules']
  },
  module: {
    preLoaders: [
      {
        test: /\.js?$/,
        loaders: ['eslint']
      },
      {
        test: /\.css$/,
        loaders: ['postcss']
      }
    ],
    loaders: [
      {
        test: /\.js$/,
        loader: 'babel',
        include: PATHS.app
      },
      {
        test: /\.html$/,
        loader: 'html',
        include: PATHS.app
      },
      {
        test: /\.json$/,
        loader: 'json',
        include: PATHS.app
      },
      {
        test: /\.woff(2)?(\?v=[0-9]\.[0-9]\.[0-9])?$/,
        loader: 'url?limit=10000&mimetype=application/font-woff',
        include: PATHS.fonts
      },
      {
        test: /\.(ttf|eot|svg)(\?v=[0-9]\.[0-9]\.[0-9])?$/,
        loader: 'file',
        include: PATHS.fonts
      },
      {
        test: /\.jpg$/,
        loader: 'file',
        include: PATHS.images
      },
      {
        test: /\.png$/,
        loader: 'url',
        include: PATHS.images
      }
    ]
  },
  postcss: parts.postcss
}

// Detect how npm is run and branch based on that
let config
switch (process.env.npm_lifecycle_event) {
  case 'build':
    config = merge(
      common,
      parts.clean(PATHS.dist),
      parts.minify(),
      parts.extractCSS(PATHS.app),
      parts.copy(PATHS.copy)
    )
    break
  default:
    config = merge(
      common,
      {
        devtool: 'eval-source-map'
      },
      parts.setupCSS(PATHS.app),
      parts.devServer({
        // Customize host/port here if needed
        host,
        port
      })
    )

    // Run opener
    opener(`http://${host}:${port}`)
}

/**
 * Webpack configuration
 */
module.exports = validate(config)